pysages/methods/improved_string.py

```diff
@@ -0,0 +1,295 @@
+# SPDX-License-Identifier: MIT
+# Copyright (c) 2020-2021: PySAGES contributors
+# See LICENSE.md and CONTRIBUTORS.md at https://github.com/SSAGESLabs/PySAGES
+
+"""
+Improved String method.
+
+The improved string methods finds the local minimum free-energy path (MFEP) trough a given landscape.
+It uses umbrella integration to advance the discretized replicas in the CV space.
+A spline interpolation helps to (equally) space the CV points along the string.
+The process converges if the MFEP is found.
+The final free-energy profile is calculated the same way as in UmbrellaIntegration.
+"""
+
+import plum
+import numpy as np
+from copy import deepcopy
+from typing import Callable, Optional, Union, Any, List
+from numpy.linalg import norm
+from scipy.interpolate import interp1d
+
+import pysages
+from pysages.methods.core import Result, SamplingMethod, _run
+from pysages.methods.harmonic_bias import HarmonicBias
+from pysages.methods.umbrella_integration import UmbrellaIntegration
+from pysages.methods.utils import HistogramLogger, listify, SerialExecutor
+from pysages.utils import dispatch
+
+
+def _test_valid_spacing(replicas, spacing):
+    if spacing is None:
+        spacing = np.diff(np.linspace(0, 1, replicas))
+    spacing = np.asarray(spacing)
+    if len(spacing) != replicas - 1:
+        RuntimeError("The provided spacing for String is not replicas - 1.")
+    if np.any(spacing <= 0):
+        RuntimeError("Provided spacing is not a positive real number monotonic increasing {space}.")
+
+    # Normalize
+    spacing /= np.sum(spacing)
+    spacing = np.asarray([0] + list(spacing))
+    spacing = np.cumsum(spacing)
+    assert abs(spacing[0]) < 1e-6
+    assert abs(spacing[-1] - 1) < 1e-6
+    spacing[0] = 0.0
+    spacing[-1] = 1.0
+
+    return spacing
+
+
+class ImprovedString(SamplingMethod):
```

Comment:
Even if we describe it as Improved String Method in the docs could we call this just `StringMethod`? Or maybe find a different name for it?

Reply:
I thought Improved String Method was a more or less standing term for a specific implementation of the general string method.
https://www.researchgate.net/profile/Eric-Vanden-Eijnden/publication/6351826_Simplified_and_Improved_String_Method_for_Computing_the_Minimum_Energy_Paths_in_Barrier_Crossing_Events/links/5523eeab0cf2c815e073df2f/Simplified-and-Improved-String-Method-for-Computing-the-Minimum-Energy-Paths-in-Barrier-Crossing-Events.pdf

In particular, the improved string method uses a spline to interpolate the path instead of using derivatives of the path without interpolation.

If we want to make it more specific. I could see `SplineString` a good alternative word.
But I don't know if anyone else has ever called the string method that.

Reply:
When I read that reference the `improved` doesn't really make me think that that is part of the name of the new method. I'd like to be more specific, and `SplineStringMethod` or `SplineString` are good names in my opinion. I'd still use in the docs and docstrings something like *This is an implementation of the improved string method proposed/described in...*

Reply:
Yeah, I know it is bad naming.
(Never name anything, "new", "improved", "novel" etc. I hate that in paper names.)

This sounds good to me.
I will change the names accordingly.
Of the class and in the doc.